src/libsyntax/parse/parser.rs

```diff
@@ -2962,13 +2962,25 @@ impl<'a> Parser<'a> {
     }
 
     /// Parse an 'if' or 'if let' expression ('if' token already eaten)
-    pub fn parse_if_expr(&mut self, attrs: ThinVec<Attribute>) -> PResult<'a, P<Expr>> {
+    pub fn parse_if_expr(&mut self, attrs: ThinVec<Attribute>,
+                         in_else: bool) -> PResult<'a, P<Expr>> {
         if self.check_keyword(keywords::Let) {
             return self.parse_if_let_expr(attrs);
         }
         let lo = self.prev_span;
         let cond = self.parse_expr_res(RESTRICTION_NO_STRUCT_LITERAL, None)?;
-        let thn = self.parse_block()?;
+        let thn = self.parse_block().map_err(|mut err| {
```

Comment:
I'm mildly worried that this error might trigger on cases where it shouldn't -- this will trigger whenever parsing the block failed, right? Even if there *was* a condition, potentially?

For example, something like this `if true else { ... }` will signal that there is a missing condition, which isn't quite right, no?

Maybe we should check whether we see the keyword `else`, instead of invoking `parse_block()`?